tests/testthat/test-return_linter.R

```diff
@@ -882,3 +882,272 @@ test_that("return_linter lints `message`, `warning` and `stopifnot`", {
     linter
   )
 })
+
+test_that("return_linter handles arbitrarily nested terminal statements", {
+  implicit_linter <- return_linter()
+  implicit_msg <- rex::rex("Use implicit return behavior; explicit return() is not needed.")
+  explicit_linter <- return_linter(return_style = "explicit")
+  explicit_msg <- rex::rex("All functions must have an explicit return().")
+
+  expect_lint(
+    trim_some("
+      foo <- function(x) {
+        if (x < 0) {
+          if (x == -1) {
+            return(TRUE)
+          }
+          if (x > -10) {
+            NA
+          } else {
+            FALSE
+          }
+        } else if (x == 0) {
+          TRUE
+        } else {
+          y <- x**2
+          if (y > 10) {
+            z <- sin(y)
+            if (z > 0) {
+              FALSE
+            } else {
+              NA
+            }
+          } else {
+            TRUE
+          }
+        }
+      }
+    "),
+    NULL,
+    implicit_linter
+  )
+
+  expect_lint(
+    trim_some("
+      foo <- function(x) {
+        if (x < 0) {
+          if (x == -1) {
+            return(TRUE)
+          }
+          if (x > -10) {
+            return(NA)
+          } else {
+            return(FALSE)
+          }
+        } else if (x == 0) {
+          return(TRUE)
+        } else {
+          y <- x**2
+          if (y > 10) {
+            z <- sin(y)
+            if (z > 0) {
+              return(FALSE)
+            } else {
+              return(NA)
+            }
+          } else {
+            return(TRUE)
+          }
+        }
+      }
+    "),
+    NULL,
+    explicit_linter
+  )
+
+  mixed_lines <- trim_some("
```

Comment:
Very nice test!

Reply:
I'm more excited we finally have a working version of this feature, it's been low-key bugging me for years that we didn't handle it properly!